js/backup.js

```diff
@@ -823,18 +1008,30 @@
       };
       return getDirectory(options);
     },
-    importFromDirectory: function(directory) {
-      var idb;
+    importFromDirectory: function(directory, options) {
+      options = options || {};
+
+      var idb, nonMessageResult;
       return openDatabase().then(function(idb_db) {
         idb = idb_db;
-        return importNonMessages(idb_db, directory);
+
+        return Promise.all([
+          loadMessagesLookup(idb_db),
+          loadConversationLookup(idb_db),
+          loadGroupsLookup(idb_db),
+        ]);
+      }).then(function(lookups) {
+        options.messageLookup = lookups[0];
+        options.conversationLookup = lookups[1];
+        options.groupLookup = lookups[2];
       }).then(function() {
-        return importConversations(idb, directory);
+        return importNonMessages(idb, directory, options);
+      }).then(function(result) {
+        nonMessageResult = result;
+        return importConversations(idb, directory, options);
       }).then(function() {
-        return directory;
-      }).then(function(path) {
         console.log('done restoring from backup!');
-        return path;
+        return nonMessageResult;
       }, function(error) {
         console.log(
           'the import went wrong:',
```

Comment:
Could we add some tests that exercise these new import modes?

Reply:
Given that import hits the disk as well as IndexedDB, we don't have the scaffolding around resetting state and so on for that kind of thing. Yes, I like the idea of better testing, even comprehensive integration tests. I recommend that we continue with manual testing for this for a while longer.

Reply:
I am ok with manual testing for parts of the UI that are hard to test, but import/export is such a critical part of our infrastructure and very much input/output driven that I’d love for us to add some automation around that to exercise various code paths and ensure we fail gracefully and don’t lose user data.